handler/push.go

```diff
@@ -119,20 +108,28 @@ func Push(
 			level.Debug(logger).Log("msg", "failed to parse text", "err", err.Error())
 			return
 		}
-		if timestampsPresent(metricFamilies) {
-			http.Error(w, "pushed metrics must not have timestamps", http.StatusBadRequest)
-			level.Debug(logger).Log("msg", "pushed metrics must not have timestamps")
-			return
-		}
 		now := time.Now()
-		addPushTimestamp(metricFamilies, now)
-		sanitizeLabels(metricFamilies, labels)
+		errCh := make(chan error, 1)
 		ms.SubmitWriteRequest(storage.WriteRequest{
 			Labels:         labels,
 			Timestamp:      now,
 			MetricFamilies: metricFamilies,
+			Replace:        replace,
+			Done:           errCh,
 		})
-		w.WriteHeader(http.StatusAccepted)
+		for err := range errCh {
+			http.Error(
```

Comment:
The documentation of [`http.Error`](https://golang.org/pkg/net/http/#Error) states "... the caller should ensure no further writes are done to w.", but it seems you're calling Error multiple times here, and therefore also write to `w` multiple times.

Reply:
Correct. I doubt that with the current code, we will ever see more than one error, but it's in principle possible (and intended to be possible) to send more than one error te tho channel. Even that will just trigger a warning from `http.Error`, but it's still something to avoid. I'll create a small follow-up PR.